### PR TITLE
[linalg] Broadcast batch for mask on sdpa lowering

### PR DESCRIPTION
**Pull request by rsuderman**
Attention often broadcasts a mask across the batch dimension as masking is usually performed the same across attention heads. Added this materialization to the mask dimensions optionally.

### CODE REVIEW

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py [hunk 1]
```diff
@@ -5513,7 +5513,7 @@ def ScaledDotProductAttentionMaskModule_basic(module, tu: TestUtils):
     query = torch.randn(2, 3, 8, 16, dtype=torch.float32)
     key = torch.randn(2, 3, 12, 16, dtype=torch.float32)
     value = torch.randn(2, 3, 12, 20, dtype=torch.float32)
-    mask = torch.randn(2, 3, 8, 12, dtype=torch.float32)
+    mask = torch.randn(2, 1, 8, 12, dtype=torch.float32)
```
Comment (dan-garvey) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py:5516]: lol smooth.

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp [hunk 2]
```diff
@@ -1727,18 +1725,78 @@ class ConvertAtenScaledDotProductAttentionOp
       mask = genericOp.getResult(0);
     }
 
+    // Broadcast the batch dimensions of the mask:
+    if (mask) {
+      auto maskTy = cast<RankedTensorType>(mask.getType());
+      int64_t rank = maskTy.getRank();
+      bool needsBroadcast = false;
+      for (int i = 0, s = rank - 2; i < s; ++i) {
+        needsBroadcast |= maskTy.getDimSize(i) != keyTy.getDimSize(i);
```
Comment (dan-garvey) [inline lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:1734]: did not know there was an or= operator. thats cool

**Review by dan-garvey — APPROVED** [review]
Looks good to me

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp (hunk 2 above)
Comment (rsuderman) [reply lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:1734]: You can go weirder with `^=` for xor equals

**Pull request merged**